### ISSUE TITLE
Continual backpropagation

### ISSUE THREAD

**Issue by tlaurie99** [tlaurie99/reinforcement_learning#1]
Use of [continual backpropagation](https://www.nature.com/articles/s41586-024-07711-7) networks to maintain plasticity of neural networks

This is done by:

1. Maintaining a contribution utility metric that is maintained for each neurons of the network
2. The neurons that contribute the least to a given state and are near 0 is recorded via index
3. If this node continues on this list for a certain number of SGD updates then it is considered "dead" in which it gets added to a pool of dead neurons
4. Every 1 / ~200 SGD updates will select one of these neurons and reinitialize it
5. Initialize the **output weights** of the neuron to 0 (so that it does not change the current output)
6. Reinitialize the **input weights** via the prior distribution (using something like [Xavier ](https://cs230.stanford.edu/section/4/) or [Kaiming](https://paperswithcode.com/method/he-initialization))

**Comment by tlaurie99** [comment]
8/29

- [ ] Begin exploration of exposing weights and keeping track of dead neurons

**Comment by tlaurie99** [comment]
Weights have been exposed, iterated through and summed over each layer. Running into issues with getting individual neuron outputs from the l layer into the l+1 layer

- [ ] Further exploration on how to expose the outputs of individual neuron outputs

> So far, it looks like a [forward hook](https://discuss.pytorch.org/t/how-can-i-extract-intermediate-layer-output-from-loaded-cnn-model/77301/3) idea is needed

**Comment by tlaurie99** [comment]
9/3 - 9/6 

Dropped using RLLIB modules and create layers from scratch using torch -- allows easy access to outputs and access to weights. Utility function is currently being calculated given a layer output (mini_batch_size, designated layer size)

- [ ] Implement threshold limit  / updating and tracking w.r.t this
- [ ] Implement weight reinitialization
- [ ] Abstract methods away into a module for easy use

**Comment by tlaurie99** [comment]
9/9 - 9/13

-> After seeing how the theory works, decided to use their [github](https://github.com/shibhansh/loss-of-plasticity/tree/main/lop) and changed the [cbp](https://github.com/shibhansh/loss-of-plasticity/blob/main/lop/algos/cbp.py) file

- [ ] See the lop architecture and see how to properly pass rllib network modules to it

- [ ] Pass networks directly to cbp

**Comment by tlaurie99** [comment]
9/30 - 10/4

Attempting to use the cbp repo as seen [here](https://github.com/shibhansh/loss-of-plasticity/tree/7bf3dfe6723a43a543fa1057a38eaf4b480f2ff3) but using with RLLIB's structure is a little odd

- [ ] Create a module / sequential / layer extractor to pass layers as a list to lop
- > the original layer extractor I had continues in the same sequence as RLLIB (i.e. SlimFC logits layer, Sequential hidden layers, followed by the SlimFC input layer) -- attempting to make a generic extractor for this in the future
